src/peep/peep.c

```diff
@@ -7772,13 +7772,285 @@ static void peep_on_enter_or_exit_ride(rct_peep *peep, int rideIndex, int flags)
+	if ((shopItem != SHOP_ITEM_MAP) && shop_item_is_souvenir(shopItem) && !has_voucher) {
+		if (((scenario_rand() & 0x7F) + 0x73) > peep->happiness)
+			return 0;
+		else if (peep->no_of_rides < 3)
+			return 0;
+	}
+
+loc_69B119:
+	if (!has_voucher) {
+		if (price != 0) {
+			if (peep->cash_in_pocket == 0) {
+				peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_SPENT_MONEY, 0xFF);
+				return 0;
+			}
+			if (price > peep->cash_in_pocket) {
+				peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_CANT_AFFORD, shopItem);
+				return 0;
+			}
+		}
+
+		///loc_69B143:
+		if (RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_TEMPERATURE, uint16) >= 21)
+			value = get_shop_hot_value(shopItem);
+		else if (RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_TEMPERATURE, uint16) <= 11)
+			value = get_shop_cold_value(shopItem);
+		else
+			value = get_shop_base_value(shopItem);
+
+		///loc_69B19D:
+		if (value < price) {
+			value -= price;
+			if (shopItem == SHOP_ITEM_UMBRELLA) {
+				if (RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_RAIN_LEVEL, sint8) != 0)
+					goto loc_69B221;
+			}
+			///loc_69B1F8:
+			value = -value;
+			if (peep->happiness >= 128)
+				value >>= 1;
+			/// loc_69B204:
+			if (peep->happiness >= 180)
+				value >>= 1;
+			/// loc 69B20D:
+			if (value > ((money16)(scenario_rand() & 0x07))) {
+				///loc_69B530:
+				// "I'm not paying that much for x"
+				uint8 thought_type = (shopItem >= 32 ? (PEEP_THOUGHT_TYPE_PHOTO2_MUCH + (shopItem - 32)) : (PEEP_THOUGHT_TYPE_BALLOON_MUCH + shopItem));
+				peep_insert_new_thought(peep, shopItem, rideIndex);
+				return 0;
+			}
+		}
+		else {
+			value -= price;
+			value = max(8, value);
+			sint8 dl = value;
+			/// loc_69B1AC
+			if ((dl >= (scenario_rand() & 0x07)) && !(RCT2_GLOBAL(RCT2_ADDRESS_PARK_FLAGS, uint32) & PARK_FLAGS_NO_MONEY)) {
+				// "This x is a really good value"
+				uint8 thought_item = (shopItem >= 32 ? (PEEP_THOUGHT_TYPE_PHOTO2 + (shopItem - 32)) : (PEEP_THOUGHT_TYPE_BALLOON + shopItem));
+				peep_insert_new_thought(peep, thought_item, rideIndex);
+			}
+			/// loc_69B1D4:
+			dl *= 4;
+			peep->happiness_growth_rate = min((peep->happiness_growth_rate + dl), 255);
+			/// loc_69B1E0:
+			peep->happiness = min((peep->happiness + dl), 255);
+		}
+	}
+
+loc_69B221:
+	if (!has_voucher) {
+		if (RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_TEMPERATURE, uint16) >= 21)
+			value = get_shop_hot_value(shopItem);
+		else if (RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_TEMPERATURE, uint16) <= 11)
+			value = get_shop_cold_value(shopItem);
+		else
+			value = get_shop_base_value(shopItem);
+
+		/// loc_69B293:
+		value -= price;
+		uint8 bl = 0;
+		if (value > -8) {
+			bl++;
+			if (value > -3) {
+				bl++;
+				if (value > 3)
+					bl++;
+			}
+		}
+		/// loc_69B2B0:
+		ride_update_satisfaction(ride, bl);
+	}
+
+///loc_69B2B7:
+	// The peep has now decided to buy the item (or, specifically, has not been
+	// dissuaded so far).
+	if (shopItem >= 32)
+		peep->item_extra_flags |= (1 << (shopItem >= 32 ? (shopItem - 32) : shopItem));
```

Comment:
The extra check on shopItem>= 32 is redundant. Same for the one below but the other way round.